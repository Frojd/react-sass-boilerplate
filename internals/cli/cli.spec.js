/* global beforeAll afterAll */

import path from 'path';
import fs from 'fs-extra';

import Cli from './cli.js';

const testFolder = path.join(process.cwd(), 'tmp_testCli');
const testConfig = {
    rootFolder: testFolder
}
const config = require('../config.js')(testConfig);

const cli = new Cli(config);
const rootFolder = path.join(
    config.rootFolder, 
    config.appFolder, 
    config.componentsFolder
)
const containerFolder = path.join(
    config.rootFolder, 
    config.appFolder, 
    config.containersFolder
)
const componentName = 'MainComponent';
const subComponentName = 'SubComponent';

describe('Test CLI functions', () => {
    beforeAll(() => {
        // Remove logging info
        console.info = () => {};
    
        // Generate essential files
        const scssFolder = path.join(config.rootFolder, config.appFolder, config.scssFolder);
        const appIndexJs = path.join(rootFolder, 'index.js');
        const appIndexScss = path.join(scssFolder, 'index.scss');
        const indexJsContent = `
export {
};`;
        fs.ensureDirSync(rootFolder);
        fs.ensureDirSync(scssFolder);
        fs.ensureDirSync(containerFolder);
        fs.writeFileSync(appIndexJs, indexJsContent);
        fs.createFileSync(appIndexScss);
    });
    
    describe('Create a classbased component', () => {
        it('Create MainComponent', () => {
            const folderPath = path.join(rootFolder, componentName);
            cli.createComponent(folderPath, componentName);
            expect(fs.existsSync(folderPath)).toBe(true);
        });

        it('Check content of MainComponent index.js', () => {
            const filePath = path.join(rootFolder, componentName, `index.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of class MainComponent.js', () => {
            const filePath = path.join(rootFolder, componentName, `${componentName}.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of MainComponent.test.js', () => {
            const filePath = path.join(rootFolder, componentName, '', `${componentName}.test.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of classbased MainComponent.scss', () => {
            const filePath = path.join(rootFolder, componentName, `${componentName}.scss`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of classbased MainComponent.json', () => {
            const filePath = path.join(rootFolder, componentName, `${componentName}.json`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });
        
        it('Should have updated app index.js', () => {
            const filePath = path.join(rootFolder, `index.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Should have updated app index.scss', () => {
            const scssFolder = path.join(config.rootFolder, config.appFolder, config.scssFolder);
            const filePath = path.join(scssFolder, `index.scss`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });
    })

    describe('Test pure based subcomponent', () => {
        beforeAll(() => {
            // Change config to pure and don't update files
            cli.config.createPure = true;
            cli.config.createClass = false;
            cli.config.updateIndexJs = false;
            cli.config.updateIndexScss = false;
        });
        
        it('Create SubComponent', () => {
            const folderPath = path.join(rootFolder, componentName, subComponentName);
            cli.createComponent(folderPath, subComponentName);
            expect(fs.existsSync(folderPath)).toBe(true);
        });
    
        it('Check content of SubComponent index.js', () => {
            const filePath = path.join(rootFolder, componentName, subComponentName, `index.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of pure SubComponent.js', () => {
            const filePath = path.join(rootFolder, componentName, subComponentName, `${subComponentName}.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of SubComponent.test.js', () => {
            const filePath = path.join(rootFolder, componentName, subComponentName, `${subComponentName}.test.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of SubComponent.scss', () => {
            const filePath = path.join(rootFolder, componentName, subComponentName, `${subComponentName}.scss`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Check content of SubComponent.json', () => {
            const filePath = path.join(rootFolder, componentName, `${componentName}.json`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Should NOT have updated app index.js', () => {
            const filePath = path.join(rootFolder, `index.js`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Should NOT have updated app index.scss', () => {
            const scssFolder = path.join(config.rootFolder, config.appFolder, config.scssFolder);
            const filePath = path.join(scssFolder, `index.scss`);
            const template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });
    })

    describe('Publish components', () => {
        it('Should create html folder and files', () => {
            const htmlPath = path.join(testFolder, config.outputPathHtmlFolder);
            cli.publishComponents(htmlPath);
            expect(fs.existsSync(htmlPath)).toBe(true);
        });

        it('Should have a listing file', () => {
            const htmlPath = path.join(testFolder, config.outputPathHtmlFolder, `index.html`);
            const template = fs.readFileSync(htmlPath, 'utf8')
            expect(template).toMatchSnapshot();
        });

        it('Should have a component file', () => {
            const htmlPath = path.join(testFolder, config.outputPathHtmlFolder, `${componentName}`, `index.html`);
            const template = fs.readFileSync(htmlPath, 'utf8')
            expect(template).toMatchSnapshot();
        });
    })

    describe('Delete components', () => {
        it('Deletes subcomponent folder and files', () => {
            const folderPath = path.join(rootFolder, componentName, subComponentName);
            cli.deleteComponent(folderPath, subComponentName);
            expect(fs.existsSync(folderPath)).toBe(false);

            const scssFolder = path.join(config.rootFolder, config.appFolder, config.scssFolder);
            let filePath = path.join(scssFolder, `index.scss`);
            let template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();

            filePath = path.join(rootFolder, `index.js`);
            template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });
    
        it('Deletes component folder and files', () => {
            const folderPath = path.join(rootFolder, componentName);
            cli.deleteComponent(folderPath, componentName);
            expect(fs.existsSync(folderPath)).toBe(false);

            const scssFolder = path.join(config.rootFolder, config.appFolder, config.scssFolder);
            let filePath = path.join(scssFolder, `index.scss`);
            let template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();

            filePath = path.join(rootFolder, `index.js`);
            template = fs.readFileSync(filePath, 'utf8')
            expect(template).toMatchSnapshot();
        });
    })

    afterAll(() => {
        cli.deleteComponentFolder(testFolder);
    });
});