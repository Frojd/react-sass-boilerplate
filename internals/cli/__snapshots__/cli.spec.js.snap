// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CLI functions Create a classbased component Check content of MainComponent index.js 1`] = `
"import MainComponent from './MainComponent';

export default MainComponent;
"
`;

exports[`Test CLI functions Create a classbased component Check content of MainComponent.test.js 1`] = `
"/**
 * Tests for MainComponent
 */

import React from 'react';
import renderer from 'react-test-renderer';
import MainComponent from './';
import data from './MainComponent.json';

describe('<MainComponent />', () => {
    it('Renders an empty MainComponent', () => {
        const componentJson = renderer
            .create(<MainComponent />)
            .toJSON();
        expect(componentJson).toBeTruthy();
    });

    it('Renders MainComponent with data', () => {
        const componentJson = renderer
            .create(<MainComponent {...data} />)
            .toJSON();
        // expect(componentJson).toMatchSnapshot();
    });
});
"
`;

exports[`Test CLI functions Create a classbased component Check content of class MainComponent.js 1`] = `
"/**
 * MainComponent
 */

import React, {PureComponent} from 'react';

class MainComponent extends PureComponent {
    render() {
        return (
            <div className=\\"MainComponent\\">
                MainComponent
            </div>
        );
    }
}

MainComponent.defaultProps = {
};

export default MainComponent;
"
`;

exports[`Test CLI functions Create a classbased component Check content of classbased MainComponent.json 1`] = `
"{}
"
`;

exports[`Test CLI functions Create a classbased component Check content of classbased MainComponent.scss 1`] = `
"/**
* MainComponent
*/

.MainComponent {
}"
`;

exports[`Test CLI functions Create a classbased component Should have updated app index.js 1`] = `
"import MainComponent from './MainComponent';

export {
    MainComponent,
};"
`;

exports[`Test CLI functions Create a classbased component Should have updated app index.scss 1`] = `
"@import '../components/MainComponent/MainComponent';
"
`;

exports[`Test CLI functions Delete components Deletes component folder and files 1`] = `""`;

exports[`Test CLI functions Delete components Deletes component folder and files 2`] = `
"
export {
};"
`;

exports[`Test CLI functions Delete components Deletes subcomponent folder and files 1`] = `
"@import '../components/MainComponent/MainComponent';
"
`;

exports[`Test CLI functions Delete components Deletes subcomponent folder and files 2`] = `
"import MainComponent from './MainComponent';

export {
    MainComponent,
};"
`;

exports[`Test CLI functions Publish components Should have a component file 1`] = `
"<!doctype html>
<html class=\\"no-js\\" lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
        <title>React testing ground</title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">

        <link rel=\\"stylesheet\\" href=\\"/devserver.css\\">
        <link rel=\\"stylesheet\\" href=\\"/static/css/index.css\\">
    </head>
    <body onload=\\"document.documentElement.classList.remove('no-js');\\">
        <div class=\\"devserver\\">
            <form class='devserver__themeform' id='devserver__menu'>
                <a href='/' class='devserver__backlink'>Back to listing</a>
                
                <select
                    class='devserver__themeselect'
                    name='componentName'
                    onChange='devFormChangeComponent(this);'
                    >
                    <option disabled>Välj komponent</option>
                    <option selected>MainComponent</option>
                </select>
            </form>
        </div>
        
        <div id=\\"root\\"><!-- content --></div>

        
        <script src=\\"/devserver.js\\"></script>
        <script src=\\"/static/js/vendor.js\\"></script>
        <script src=\\"/static/js/index.js\\"></script>
    </body>
</html>
"
`;

exports[`Test CLI functions Publish components Should have a listing file 1`] = `
"<!doctype html>
<html class=\\"no-js\\" lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
        <title>React testing ground</title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">

        <link rel=\\"stylesheet\\" href=\\"/devserver.css\\">
        <link rel=\\"stylesheet\\" href=\\"/static/css/index.css\\">
    </head>
    <body onload=\\"document.documentElement.classList.remove('no-js');\\">
        <div class=\\"devserver\\">
            <form class='devserver__themeform' id='devserver__menu'>
                <a href='/' class='devserver__backlink'>Back to listing</a>
                
                <select
                    class='devserver__themeselect'
                    name='componentName'
                    onChange='devFormChangeComponent(this);'
                    >
                    <option disabled>Välj komponent</option>
                    <option >MainComponent</option>
                </select>
            </form>
        </div>
        
        <a href=\\"/MainComponent\\">MainComponent</a><br/>

        
        <script src=\\"/devserver.js\\"></script>
        <script src=\\"/static/js/vendor.js\\"></script>
        <script src=\\"/static/js/index.js\\"></script>
    </body>
</html>
"
`;

exports[`Test CLI functions Test pure based subcomponent Check content of SubComponent index.js 1`] = `
"import SubComponent from './SubComponent';

export default SubComponent;
"
`;

exports[`Test CLI functions Test pure based subcomponent Check content of SubComponent.json 1`] = `
"{}
"
`;

exports[`Test CLI functions Test pure based subcomponent Check content of SubComponent.scss 1`] = `
"/**
* SubComponent
*/

.SubComponent {
}"
`;

exports[`Test CLI functions Test pure based subcomponent Check content of SubComponent.test.js 1`] = `
"/**
 * Tests for SubComponent
 */

import React from 'react';
import renderer from 'react-test-renderer';
import SubComponent from './';
import data from './SubComponent.json';

describe('<SubComponent />', () => {
    it('Renders an empty SubComponent', () => {
        const componentJson = renderer
            .create(<SubComponent />)
            .toJSON();
        expect(componentJson).toBeTruthy();
    });

    it('Renders SubComponent with data', () => {
        const componentJson = renderer
            .create(<SubComponent {...data} />)
            .toJSON();
        // expect(componentJson).toMatchSnapshot();
    });
});
"
`;

exports[`Test CLI functions Test pure based subcomponent Check content of pure SubComponent.js 1`] = `
"/**
 * SubComponent
 */

import React, {PureComponent} from 'react';

class SubComponent extends PureComponent {
    render() {
        return (
            <div className=\\"SubComponent\\">
                SubComponent
            </div>
        );
    }
}

SubComponent.defaultProps = {
};

export default SubComponent;
"
`;

exports[`Test CLI functions Test pure based subcomponent Should NOT have updated app index.js 1`] = `
"import MainComponent from './MainComponent';

export {
    MainComponent,
};"
`;

exports[`Test CLI functions Test pure based subcomponent Should NOT have updated app index.scss 1`] = `
"@import '../components/MainComponent/MainComponent';
"
`;
